Mathlib/GroupTheory/RegularWreathProduct.lean

```diff
@@ -0,0 +1,301 @@
+
+@[simp]
+theorem mul_left (a b : D ≀ᵣ Q) : (a * b).1 = a.1 * (((fun x => b.1 (a.2⁻¹ * x)))) := rfl
+
+@[simp]
+theorem mul_right (a b : D ≀ᵣ Q) : (a * b).right = a.right * b.right := rfl
+
+instance : One (RegularWreathProduct D Q) where one := ⟨1, 1⟩
+
+@[simp]
+theorem one_left : (1 : D ≀ᵣ Q).left = 1 := rfl
+
+@[simp]
+theorem one_right : (1 : D ≀ᵣ Q).right = 1 := rfl
+
+instance : Inv (RegularWreathProduct D Q) where
+  inv x := ⟨((fun k => x.1⁻¹ (x.2 * k))), x.2⁻¹⟩
+
+@[simp]
+theorem inv_left (a : D ≀ᵣ Q) :
+    a⁻¹.left = ((fun x => a.left⁻¹ (a.right * x))) := rfl
+
+@[simp]
+theorem inv_right (a : D ≀ᵣ Q) : a⁻¹.right = a.right⁻¹ := rfl
+
+instance : Group (RegularWreathProduct D Q) where
+  mul_assoc a b c := by ext <;> simp [mul_assoc]
+  one_mul a := by ext <;> simp
+  mul_one a := by ext <;> simp
+  inv_mul_cancel a := by ext <;> simp
+
+instance : Inhabited (RegularWreathProduct D Q) := ⟨1⟩
+
+/-- The canonical projection map `D ≀ᵣ Q →* Q`, as a group hom. -/
+def rightHom : D ≀ᵣ Q →* Q where
+  toFun := RegularWreathProduct.right
+  map_one' := rfl
+  map_mul' _ _ := rfl
+
+/-- The canonical map `Q →* D ≀ᵣ Q` sending `q` to `⟨1, q⟩` -/
+def inl : Q →* D ≀ᵣ Q where
+  toFun q := ⟨1, q⟩
+  map_one' := rfl
+  map_mul' _ _ := by ext <;> simp
+
+@[simp]
+theorem left_inl (q : Q) : (inl q : D ≀ᵣ Q).left = 1 := rfl
+
+@[simp]
+theorem right_inl (q : Q) : (inl q : D ≀ᵣ Q).right = q := rfl
+
+@[simp]
+theorem rightHom_eq_right : (rightHom : D ≀ᵣ Q → Q) = right := rfl
+
+@[simp]
+theorem rightHom_comp_inl_eq_id : (rightHom : D ≀ᵣ Q →* Q).comp inl = MonoidHom.id _ := by ext; simp
+
+@[simp]
+theorem fun_id (q : Q) : rightHom (inl q : D ≀ᵣ Q) = q := by simp
+
+/-- The equivalence map for the representation as a product. -/
+def equivProd D Q : D ≀ᵣ Q ≃ (Q → D) × Q where
+  toFun := fun ⟨d, q⟩ => ⟨d, q⟩
+  invFun := fun ⟨d, q⟩ => ⟨d, q⟩
+  left_inv := fun _ => rfl
+  right_inv := fun _ => rfl
+
+omit [Group D] [Group Q] in
+lemma equivProdInj : Function.Injective (equivProd D Q).toFun := by
+  intro a b; simp
+
+instance [Finite D] [Finite Q] : Finite (D ≀ᵣ Q) :=
+  Finite.of_equiv _ (equivProd D Q).symm
+
+omit [Group D] [Group Q] in
+theorem card [Finite Q] : Nat.card (D ≀ᵣ Q) = Nat.card D ^ Nat.card Q * Nat.card Q := by
+  rw [Nat.card_congr (equivProd D Q), Nat.card_prod (Q → D) Q, Nat.card_fun]
+
+/-- Define an isomorphism from `D₁ ≀ᵣ Q₁` to `D₂ ≀ᵣ Q₂`
+given isomorphisms `D₁ ≀ᵣ Q₁` and `Q₁ ≃* Q₂`. -/
+def congr {D₁ Q₁ D₂ Q₂ : Type*} [Group D₁] [Group Q₁] [Group D₂] [Group Q₂]
+    (f : D₁ ≃* D₂) (g : Q₁ ≃* Q₂) :
+    D₁ ≀ᵣ Q₁ ≃* D₂ ≀ᵣ Q₂ where
+  toFun x := ⟨f ∘ (x.left ∘ g.symm), g x.right⟩
+  invFun x := ⟨(f.symm ∘ x.left) ∘ g, g.symm x.right⟩
+  left_inv x := by ext <;> simp
+  right_inv x := by ext <;> simp
+  map_mul' x y := by ext <;> simp
+
+section perm
+
+variable (D Q) (Λ : Type*) [MulAction D Λ]
+
+instance instSMulRWP : SMul (D ≀ᵣ Q) (Λ × Q) where
```

Comment:
I suggest either giving this a more descriptive name or dropping the name altogether.